### PR TITLE
gh-118272: Clear generator frame's locals when the generator is closed

### LINKED ISSUE #118272: Generator .close does not release resources
# Bug report

### Bug description:

Sending a generator `.close()` in Python 3.12 does not release resources used by local variables as it used to do in earlier versions of Python. Wrapping the generator content in a `try: ... except GeneratorExit: pass` releases the resources as expected.

The following shows the difference in space usage between wrapping the code with `try`-`except` and not.

```python
def generator_using_a_lot_of_space():
    a_big_set = set(range(10_000_000))
    yield 42

g = generator_using_a_lot_of_space()
input('A <press enter>')  # 13.4 MB
print(next(g))
input('B <press enter>')  # 576.6 MB
g.close()
input('C <press enter>')  # 576.6 MB  <== space usage used to drop to 13-14 MB

def generator_using_a_lot_of_space():
    a_big_set = set(range(10_000_000))
    try: 
        yield 42
    except GeneratorExit:
        pass
    
g = generator_using_a_lot_of_space()
input('D <press enter>')  # 14.6 MB
print(next(g))
input('E <press enter>')  # 575.6 MB
g.close()
input('F <press enter>')  # 14.6 MB  <== drop in space usage as expected
```

The above space measurements were done with 
Python 3.12.3 (tags/v3.12.3:f6650f9, Apr  9 2024, 14:05:25) [MSC v.1938 64 bit (AMD64)] on win32
and looking at the space usage reported by the Windows 11 Task Manager.

### CPython versions tested on:

3.12

### Operating systems tested on:

Windows

### Linked PRs
* gh-118277
* gh-118451
* gh-118478

### PR DESCRIPTION
**Pull request by iritkatriel**
Fixes #118272.

Co-authored-by: Thomas Grainger <tagrain@gmail.com>

* Issue: gh-118272

### CODE REVIEW

File: Lib/test/test_generators.py [hunk 1]
```diff
@@ -532,6 +532,26 @@ def f():
...
+
+        def genfn():
+            a = f
+            yield
+
+        g = genfn()
+        next(g)
+        del f
+        g.close()
+        gc.collect()
```
Comment (graingert) [inline Lib/test/test_generators.py]:
```suggestion
        support.gc_collect()
```

**Review by gvanrossum — APPROVED** [review]
Yeah, LGTM!

File: Python/frame.c [hunk 2]
```diff
@@ -94,6 +94,17 @@ take_ownership(PyFrameObject *f, _PyInterpreterFrame *frame)
     }
 }
 
+void
+_PyFrame_ClearLocals(_PyInterpreterFrame *frame)
+{
+    assert(frame->stacktop >= 0);
+    for (int i = 0; i < frame->stacktop; i++) {
```
Comment (markshannon) [inline Python/frame.c:101]: I think `frame->stacktop = 0;` needs to come before the loop to prevent the GC seeing freed objects.

Comment (gvanrossum) [reply Python/frame.c:101]: ~~Let me make a fix for that.~~ NM, approved gh-118478 instead.

**Pull request merged**